cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,321 @@
+
+  // TODO: (CDAP-8439) determine the appropriate size
+  private static final int FETCH_SIZE = 1024 * 1024;        // Use a default fetch size.
+  private static final int SO_TIMEOUT = 5 * 1000;           // 5 seconds.
+  private static final long LATEST = kafka.api.OffsetRequest.LatestTime();
+  private static final long EARLIEST = kafka.api.OffsetRequest.EarliestTime();
+  // A rough estimation of Kafka publishing delay of a message from the original cluster and the destination cluster
+  // during cross-cluster replication
+  private static final long REPLICATION_DELAY = 5 * 60 * 1000;
+  // The max mount of time that the log event time of a message with smaller offset
+  // can exceed a message with larger offset
+  private static final long MAX_TIME_GAP = 60 * 1000; // 1 min
+
+  private final BrokerService brokerService;
+  private final String topic;
+  private final LoggingEventSerializer serializer;
+
+  public KafkaOffsetResolver(BrokerService brokerService, String topic) {
+    this.brokerService = brokerService;
+    this.topic = topic;
+    this.serializer = new LoggingEventSerializer();
+  }
+
+  /**
+   * Check whether the message fetched with the offset {@code checkpoint.getNextOffset() - 1} contains the
+   * same timestamp as in the given checkpoint. If they match, directly return {@code checkpoint.getNextOffset()}.
+   * If they don't, search for the smallest offset of the message with the same log event time
+   * as {@code checkpoint.getNextEventTime()}
+   *
+   * @param checkpoint A {@link Checkpoint} containing the next offset of a message and its log event timestamp.
+   * @param partition the partition in the topic for searching matching offset
+   * @return the next offset of the message with smallest offset and log event time equal to
+   *         {@code checkpoint.getNextEventTime()}.
+   *         {@code -1} if no such offset can be found or {@code checkpoint.getNextOffset()} is negative.
+   */
+  public long getMatchingOffset(final Checkpoint checkpoint, final int partition) {
+    if (checkpoint.getNextOffset() < 0) {
+      return -1;
+    }
+
+    // Get BrokerInfo for constructing SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      return EARLIEST; // Return earliest offset if brokerInfo is null
+    }
+    SimpleConsumer simpleConsumer
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT, FETCH_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    long timeStamp = getEventTimeByOffset(simpleConsumer, partition, clientId, checkpoint.getNextOffset() - 1);
```

Comment:
I don't think `checkpoint.getNextOffset() == 0` is a valid checkpoint because checkpoints are constructed with `message.nextOffset()`, but no message has nextOffset 0.